File: routes/booking.py
# ============================================================
# MILLROD SWIM ACADEMY
# BOOKING ROUTES
# ============================================================

from flask import (
    Blueprint,
    request,
    jsonify,
    render_template,
)

from datetime import datetime
import secrets

from config import Config
from database import get_db_connection

import email_service


# ============================================================
# BLUEPRINT
# ============================================================

booking_bp = Blueprint(
    "booking",
    __name__,
    url_prefix="/api",
)


# ============================================================
# GENERATE APPROVAL TOKEN
# ============================================================

def generate_approval_token():
    return secrets.token_urlsafe(32)


# ============================================================
# CONVERT DATABASE ROW TO DICTIONARY
# ============================================================

def row_to_dict(cursor, row):

    if row is None:
        return None

    if hasattr(row, "keys"):
        return dict(row)

    columns = [
        description[0]
        for description in cursor.description
    ]

    return dict(
        zip(
            columns,
            row,
        )
    )


# ============================================================
# CREATE BOOKING
# ============================================================

@booking_bp.route(
    "/create-booking",
    methods=["POST"],
)
def create_booking():

    conn = None

    try:

        data = request.get_json(
            silent=True
        ) or {}

        print("========================================")
        print("NEW BOOKING REQUEST")
        print("BOOKING DATA:")
        print(data)
        print("========================================")

        # ----------------------------------------------------
        # CUSTOMER
        # ----------------------------------------------------

        student_name = (
            data.get("student_name")
            or data.get("name")
            or ""
        ).strip()

        parent_name = (
            data.get("parent_name")
            or ""
        ).strip()

        email = (
            data.get("email")
            or ""
        ).strip()

        phone = (
            data.get("phone")
            or ""
        ).strip()

        # ----------------------------------------------------
        # LESSON
        # ----------------------------------------------------

        lesson_type = (
            data.get("lesson_type")
            or ""
        ).strip()

        package = (
            data.get("package")
            or ""
        ).strip()

        lesson_date = (
            data.get("lesson_date")
            or data.get("date")
            or ""
        ).strip()

        lesson_time = (
            data.get("lesson_time")
            or data.get("time")
            or ""
        ).strip()

        # ----------------------------------------------------
        # OTHER INFORMATION
        # ----------------------------------------------------

        dob = (
            data.get("dob")
            or ""
        ).strip()

        age = data.get("age") or ""

        emergency_contact = (
            data.get("emergency_contact")
            or ""
        ).strip()

        emergency_phone = (
            data.get("emergency_phone")
            or ""
        ).strip()

        swimming_experience = (
            data.get("swimming_experience")
            or ""
        ).strip()

        medical = (
            data.get("medical")
            or ""
        ).strip()

        notes = (
            data.get("notes")
            or ""
        ).strip()

        price = (
            data.get("price")
            or ""
        ).strip()

        # ----------------------------------------------------
        # VALIDATION
        # ----------------------------------------------------

        if not student_name:
            return jsonify({
                "success": False,
                "error": "Student name is required.",
            }), 400

        if not email:
            return jsonify({
                "success": False,
                "error": "Email address is required.",
            }), 400

        if not phone:
            return jsonify({
                "success": False,
                "error": "Phone number is required.",
            }), 400

        if not lesson_type:
            return jsonify({
                "success": False,
                "error": "Lesson type is required.",
            }), 400

        if not package:
            return jsonify({
                "success": False,
                "error": "Package is required.",
            }), 400

        if not lesson_date:
            return jsonify({
                "success": False,
                "error": "Lesson date is required.",
            }), 400

        if not lesson_time:
            return jsonify({
                "success": False,
                "error": "Lesson time is required.",
            }), 400

        # ----------------------------------------------------
        # DATABASE
        # ----------------------------------------------------

        conn = get_db_connection()
        cursor = conn.cursor()

        # ----------------------------------------------------
        # CHECK EXISTING BOOKING
        # ----------------------------------------------------

        cursor.execute(
            """
            SELECT id
            FROM bookings
            WHERE lesson_date = %s
              AND lesson_time = %s
              AND status != 'cancelled'
            LIMIT 1
            """,
            (
                lesson_date,
                lesson_time,
            ),
        )

        existing_booking = cursor.fetchone()

        if existing_booking:

            return jsonify({
                "success": False,
                "error": (
                    "This lesson time is already booked. "
                    "Please choose another time."
                ),
            }), 409

        # ----------------------------------------------------
        # APPROVAL TOKEN
        # ----------------------------------------------------

        approval_token = (
            generate_approval_token()
        )

        # ----------------------------------------------------
        # CREATE BOOKING
        # ----------------------------------------------------

        created_at = datetime.now()

        cursor.execute(
            """
            INSERT INTO bookings (
                name,
                email,
                phone,
                lesson_type,
                package,
                lesson_date,
                lesson_time,
                payment_status,
                stripe_payment_id,
                status,
                reminder_sent,
                approval_token,
                created_at
            )
            VALUES (
                %s,
                %s,
                %s,
                %s,
                %s,
                %s,
                %s,
                %s,
                %s,
                %s,
                %s,
                %s,
                %s
            )
            RETURNING id
            """,
            (
                student_name,
                email,
                phone,
                lesson_type,
                package,
                lesson_date,
                lesson_time,
                "pending",
                None,
                "pending",
                False,
                approval_token,
                created_at,
            ),
        )

        booking_id = cursor.fetchone()[0]

        conn.commit()

        # ----------------------------------------------------
        # BOOKING DICTIONARY
        # ----------------------------------------------------

        booking = {
            "id": booking_id,

            "name": student_name,

            "student_name": student_name,

            "parent_name": parent_name,

            "email": email,

            "phone": phone,

            "dob": dob,

            "age": age,

            "emergency_contact":
                emergency_contact,

            "emergency_phone":
                emergency_phone,

            "swimming_experience":
                swimming_experience,

            "lesson_type":
                lesson_type,

            "package":
                package,

            "lesson_date":
                lesson_date,

            "lesson_time":
                lesson_time,

            "price":
                price,

            "medical":
                medical,

            "notes":
                notes,

            "payment_status":
                "pending",

            "status":
                "pending",

            "approval_token":
                approval_token,

            "created_at":
                created_at,
        }

        print("========================================")
        print("BOOKING CREATED")
        print(
            "BOOKING ID:",
            booking_id,
        )
        print(
            "APPROVAL TOKEN PRESENT:",
            bool(approval_token),
        )
        print("========================================")

        # ----------------------------------------------------
        # SEND OWNER EMAIL
        # ----------------------------------------------------

        try:

            if not Config.OWNER_EMAIL:
                raise RuntimeError(
                    "OWNER_EMAIL is not configured."
                )

            if not Config.RESEND_API_KEY:
                raise RuntimeError(
                    "RESEND_API_KEY is not configured."
                )

            if not Config.EMAIL_FROM:
                raise RuntimeError(
                    "EMAIL_FROM is not configured."
                )

            print(
                "SENDING OWNER APPROVAL EMAIL..."
            )

            email_result = (
                email_service
                .send_admin_notification(
                    booking
                )
            )

            print(
                "OWNER EMAIL RESULT:",
                repr(email_result),
            )

            if not email_result:
                raise RuntimeError(
                    "send_admin_notification() "
                    "returned False."
                )

            print(
                f"OWNER APPROVAL EMAIL SENT "
                f"FOR BOOKING #{booking_id}"
            )

        except Exception as email_error:

            print(
                "========================================"
            )

            print(
                "OWNER EMAIL ERROR:"
            )

            print(
                repr(email_error)
            )

            print(
                "========================================"
            )

            return jsonify({
                "success": False,
                "booking_id": booking_id,
                "error": (
                    "Your booking was saved, "
                    "but the owner approval email "
                    "could not be sent."
                ),
            }), 500

        # ----------------------------------------------------
        # SUCCESS
        # ----------------------------------------------------

        return jsonify({
            "success": True,
            "booking_id": booking_id,
            "message": (
                "Booking request sent for approval."
            ),
        }), 201

    except Exception as error:

        if conn:

            try:
                conn.rollback()
            except Exception:
                pass

        print(
            "CREATE BOOKING ERROR:",
            repr(error),
        )

        return jsonify({
            "success": False,
            "error": (
                "Unable to create your booking. "
                "Please try again."
            ),
        }), 500

    finally:

        if conn:
            conn.close()


# ============================================================
# APPROVE BOOKING
# ============================================================

@booking_bp.route(
    "/approve-booking",
    methods=["GET"],
)
def approve_booking():

    conn = None

    try:

        token = (
            request.args.get(
                "token",
                "",
            ).strip()
        )

        booking_id = (
            request.args.get(
                "booking_id",
                "",
            ).strip()
        )

        print("========================================")
        print("APPROVE BOOKING REQUEST")
        print(
            "TOKEN PRESENT:",
            bool(token),
        )
        print(
            "BOOKING ID:",
            booking_id,
        )
        print("========================================")

        conn = get_db_connection()
        cursor = conn.cursor()

        # ----------------------------------------------------
        # FIND BY SECURE TOKEN
        # ----------------------------------------------------

        if token:

            cursor.execute(
                """
                SELECT *
                FROM bookings
                WHERE approval_token = %s
                LIMIT 1
                """,
                (token,),
            )

        # ----------------------------------------------------
        # BACKWARD COMPATIBILITY
        # ----------------------------------------------------

        elif booking_id:

            if not booking_id.isdigit():

                return (
                    "Invalid booking_id",
                    400,
                )

            cursor.execute(
                """
                SELECT *
                FROM bookings
                WHERE id = %s
                LIMIT 1
                """,
                (int(booking_id),),
            )

        else:

            return (
                "Invalid approval link",
                400,
            )

        booking_row = cursor.fetchone()

        if not booking_row:

            return (
                "Booking not found",
                404,
            )

        booking = row_to_dict(
            cursor,
            booking_row,
        )

        print(
            "FOUND BOOKING:",
            booking.get("id"),
        )

        # ----------------------------------------------------
        # ALREADY CONFIRMED
        # ----------------------------------------------------

        if (
            booking.get("status")
            == "confirmed"
        ):

            return render_template(
                "approval_result.html",
                success=True,
                title="Booking Already Confirmed",
                message=(
                    "This booking has already "
                    "been approved."
                ),
            )

        # ----------------------------------------------------
        # CANCELLED
        # ----------------------------------------------------

        if (
            booking.get("status")
            == "cancelled"
        ):

            return render_template(
                "approval_result.html",
                success=False,
                title="Booking Cancelled",
                message=(
                    "This booking has already "
                    "been cancelled."
                ),
            )

        # ----------------------------------------------------
        # CONFIRM BOOKING
        # ----------------------------------------------------

        cursor.execute(
            """
            UPDATE bookings
            SET status = 'confirmed'
            WHERE id = %s
            """,
            (booking["id"],),
        )

        conn.commit()

        booking["status"] = "confirmed"

        print(
            f"BOOKING #{booking['id']} "
            "CONFIRMED"
        )

        # ----------------------------------------------------
        # CUSTOMER APPROVAL EMAIL
        # ----------------------------------------------------

        try:

            if hasattr(
                email_service,
                "send_booking_approved",
            ):

                email_result = (
                    email_service
                    .send_booking_approved(
                        booking
                    )
                )

                print(
                    "CUSTOMER APPROVAL EMAIL RESULT:",
                    repr(email_result),
                )

            else:

                print(
                    "WARNING: "
                    "send_booking_approved() "
                    "does not exist."
                )

        except Exception as email_error:

            print(
                "CUSTOMER APPROVAL EMAIL ERROR:",
                repr(email_error),
            )

        # ----------------------------------------------------
        # SUCCESS PAGE
        # ----------------------------------------------------

        return render_template(
            "approval_result.html",
            success=True,
            title="Booking Approved!",
            message=(
                f"Booking #{booking['id']} "
                "has been successfully approved. "
                "The customer has been notified."
            ),
        )

    except Exception as error:

        if conn:

            try:
                conn.rollback()
            except Exception:
                pass

        print(
            "APPROVE BOOKING ERROR:",
            repr(error),
        )

        return render_template(
            "approval_result.html",
            success=False,
            title="Approval Error",
            message=(
                "We could not approve this booking. "
                "Please try again."
            ),
        ), 500

    finally:

        if conn:
            conn.close()


# ============================================================
# REJECT BOOKING
# ============================================================

@booking_bp.route(
    "/reject-booking",
    methods=["GET"],
)
def reject_booking():

    conn = None

    try:

        token = (
            request.args.get(
                "token",
                "",
            ).strip()
        )

        booking_id = (
            request.args.get(
                "booking_id",
                "",
            ).strip()
        )

        conn = get_db_connection()
        cursor = conn.cursor()

        # ----------------------------------------------------
        # FIND BY TOKEN
        # ----------------------------------------------------

        if token:

            cursor.execute(
                """
                SELECT *
                FROM bookings
                WHERE approval_token = %s
                LIMIT 1
                """,
                (token,),
            )

        # ----------------------------------------------------
        # OLD BOOKING ID
        # ----------------------------------------------------

        elif booking_id:

            if not booking_id.isdigit():

                return (
                    "Invalid booking_id",
                    400,
                )

            cursor.execute(
                """
                SELECT *
                FROM bookings
                WHERE id = %s
                LIMIT 1
                """,
                (int(booking_id),),
            )

        else:

            return (
                "Invalid approval link",
                400,
            )

        booking_row = cursor.fetchone()

        if not booking_row:

            return (
                "Booking not found",
                404,
            )

        booking = row_to_dict(
            cursor,
            booking_row,
        )

        # ----------------------------------------------------
        # ALREADY CANCELLED
        # ----------------------------------------------------

        if (
            booking.get("status")
            == "cancelled"
        ):

            return render_template(
                "approval_result.html",
                success=False,
                title="Booking Already Cancelled",
                message=(
                    "This booking has already "
                    "been cancelled."
                ),
            )

        # ----------------------------------------------------
        # CANCEL BOOKING
        # ----------------------------------------------------

        cursor.execute(
            """
            UPDATE bookings
            SET status = 'cancelled'
            WHERE id = %s
            """,
            (booking["id"],),
        )

        conn.commit()

        booking["status"] = "cancelled"

        print(
            f"BOOKING #{booking['id']} "
            "CANCELLED"
        )

        # ----------------------------------------------------
        # CUSTOMER REJECTION EMAIL
        # ----------------------------------------------------

        try:

            if hasattr(
                email_service,
                "send_booking_rejected",
            ):

                email_result = (
                    email_service
                    .send_booking_rejected(
                        booking
                    )
                )

                print(
                    "CUSTOMER REJECTION EMAIL RESULT:",
                    repr(email_result),
                )

            else:

                print(
                    "WARNING: "
                    "send_booking_rejected() "
                    "does not exist."
                )

        except Exception as email_error:

            print(
                "CUSTOMER REJECTION EMAIL ERROR:",
                repr(email_error),
            )

        # ----------------------------------------------------
        # SUCCESS PAGE
        # ----------------------------------------------------

        return render_template(
            "approval_result.html",
            success=True,
            title="Booking Cancelled",
            message=(
                f"Booking #{booking['id']} "
                "has been cancelled. "
                "The customer has been notified."
            ),
        )

    except Exception as error:

        if conn:

            try:
                conn.rollback()
            except Exception:
                pass

        print(
            "REJECT BOOKING ERROR:",
            repr(error),
        )

        return render_template(
            "approval_result.html",
            success=False,
            title="Cancellation Error",
            message=(
                "We could not cancel this booking. "
                "Please try again."
            ),
        ), 500

    finally:

        if conn:
            conn.close()


# ============================================================
# BOOKED SLOTS
# ============================================================

@booking_bp.route(
    "/booked-slots",
    methods=["GET"],
)
def booked_slots():

    conn = None

    try:

        conn = get_db_connection()
        cursor = conn.cursor()

        cursor.execute(
            """
            SELECT
                lesson_date,
                lesson_time
            FROM bookings
            WHERE status != 'cancelled'
            ORDER BY lesson_date, lesson_time
            """
        )

        rows = cursor.fetchall()

        slots = []

        for row in rows:

            item = row_to_dict(
                cursor,
                row,
            )

            slots.append({
                "date": str(
                    item.get(
                        "lesson_date"
                    )
                ),

                "time": str(
                    item.get(
                        "lesson_time"
                    )
                ),
            })

        return jsonify(slots)

    except Exception as error:

        print(
            "BOOKED SLOTS ERROR:",
            repr(error),
        )

        return jsonify({
            "success": False,
            "error":
                "Unable to load booked slots.",
        }), 500

    finally:

        if conn:
            conn.close()

File: static/booking.js
/* =========================================================
   MILLROD SWIM ACADEMY - PROFESSIONAL BOOKING SCRIPT
========================================================= */

document.addEventListener("DOMContentLoaded", () => {
  "use strict";

  const API = {
    createBooking: "/api/create-booking",
    bookings: "/api/bookings",
    bookedSlots: "/api/booked-slots",
  };

  // Keep synchronized with Config.LESSON_PRICES.
  const lessonPrices = {
    "Private Lesson": {
      "Single Lesson": 80,
      "4 Lessons Package": 300,
      "8 Lessons Package": 560,
      "Monthly Program": 1000,
    },

    "Semi-Private Lesson": {
      "Single Lesson": 120,
      "4 Lessons Package": 450,
      "8 Lessons Package": 850,
      "Monthly Program": 1500,
    },

    "Group Lesson": {
      "Single Lesson": 60,
      "4 Lessons Package": 220,
      "8 Lessons Package": 400,
      "Monthly Program": 700,
    },
  };

  const timeSlots = [
    ["09:00", "09:00 AM"],
    ["10:00", "10:00 AM"],
    ["11:00", "11:00 AM"],
    ["12:00", "12:00 PM"],
    ["13:00", "01:00 PM"],
    ["14:00", "02:00 PM"],
    ["15:00", "03:00 PM"],
    ["16:00", "04:00 PM"],
    ["17:00", "05:00 PM"],
  ];

  const form = document.getElementById("bookingForm");
  const calendarElement = document.getElementById("calendar");

  const lessonType = document.getElementById("lesson_type");
  const packageType = document.getElementById("package");

  const dateInput = document.getElementById("lesson_date");
  const timeSelect = document.getElementById("lesson_time");

  const priceDisplay = document.getElementById("priceDisplay");

  const sendApprovalBtn = document.getElementById("sendApprovalBtn");

  const bookingMessage = document.getElementById("bookingMessage");

  const agreement = document.getElementById("registrationAgreement");

  const successModal = document.getElementById("successModal");

  const closeSuccessModal = document.getElementById("closeSuccessModal");

  const successDoneBtn = document.getElementById("successDoneBtn");

  let calendar = null;
  let bookedSlots = new Set();
  let isSubmitting = false;

  if (!form) {
    console.error("bookingForm was not found.");
    return;
  }

  /* =========================================================
     MESSAGES
  ========================================================= */

  function showMessage(message, type = "info") {
    if (!bookingMessage) return;

    bookingMessage.textContent = message;
    bookingMessage.className = `booking-message ${type}`;
  }

  function clearMessage() {
    if (!bookingMessage) return;

    bookingMessage.textContent = "";
    bookingMessage.className = "booking-message";
  }

  /* =========================================================
     LOADING STATE
  ========================================================= */

  function setLoading(loading) {
    isSubmitting = loading;

    if (!sendApprovalBtn) return;

    sendApprovalBtn.disabled = loading;

    const text = sendApprovalBtn.querySelector(".button-text");

    const spinner = sendApprovalBtn.querySelector(".button-loading");

    if (text) {
      text.hidden = loading;
    }

    if (spinner) {
      spinner.hidden = !loading;
    }
  }

  /* =========================================================
     HELPERS
  ========================================================= */

  function value(id) {
    const element = document.getElementById(id);

    return element ? String(element.value || "").trim() : "";
  }

  function normalizeDate(value) {
    return value ? String(value).trim().slice(0, 10) : "";
  }

  function normalizeTime(value) {
    if (!value) return "";

    const match = String(value)
      .trim()
      .match(/^(\d{1,2}):(\d{2})/);

    return match
      ? `${match[1].padStart(2, "0")}:${match[2]}`
      : String(value).trim();
  }

  function slotKey(date, time) {
    return `${normalizeDate(date)}|` + `${normalizeTime(time)}`;
  }

  function isSunday(date) {
    const d = new Date(`${date}T12:00:00`);

    return !Number.isNaN(d.getTime()) && d.getDay() === 0;
  }

  /* =========================================================
     PRICE
  ========================================================= */

  function getSelectedPrice() {
    return lessonPrices[lessonType?.value]?.[packageType?.value] ?? null;
  }

  function updatePrice() {
    if (!priceDisplay) return;

    const price = getSelectedPrice();

    if (price === null) {
      priceDisplay.textContent = "Select a lesson and package";
      return;
    }

    priceDisplay.textContent = `$${price.toLocaleString("en-US", {
      minimumFractionDigits: 2,
      maximumFractionDigits: 2,
    })}`;
  }

  lessonType?.addEventListener("change", updatePrice);

  packageType?.addEventListener("change", updatePrice);

  /* =========================================================
     LOAD BOOKED SLOTS
  ========================================================= */

  async function loadBookedSlots() {
    try {
      const response = await fetch(API.bookedSlots, {
        headers: {
          Accept: "application/json",
        },
        cache: "no-store",
      });

      if (!response.ok) {
        throw new Error(`HTTP ${response.status}`);
      }

      const data = await response.json();

      if (!Array.isArray(data)) {
        throw new Error("Invalid availability response");
      }

      bookedSlots = new Set();

      data.forEach((slot) => {
        if (!slot) return;

        let date = normalizeDate(slot.date || slot.lesson_date);

        let time = normalizeTime(slot.time || slot.lesson_time);

        if ((!date || !time) && slot.start) {
          const parts = String(slot.start).split("T");

          date = normalizeDate(parts[0]);

          time = normalizeTime(parts[1]);
        }

        if (date && time) {
          bookedSlots.add(slotKey(date, time));
        }
      });

      updateTimeOptions();

      return true;
    } catch (error) {
      console.error("Availability error:", error);

      showMessage(
        "Availability could not be refreshed. Please try again.",
        "info",
      );

      return false;
    }
  }

  /* =========================================================
     TIME SELECT
  ========================================================= */

  function resetTimeSelect(message = "Select a date first") {
    if (!timeSelect) return;

    timeSelect.innerHTML = "";

    const option = document.createElement("option");

    option.value = "";
    option.textContent = message;
    option.selected = true;

    timeSelect.appendChild(option);

    timeSelect.disabled = true;
  }

  function updateTimeOptions() {
    if (!timeSelect) return;

    const date = normalizeDate(dateInput?.value);

    if (!date) {
      resetTimeSelect();
      return;
    }

    if (isSunday(date)) {
      resetTimeSelect("Closed on Sundays");
      return;
    }

    timeSelect.innerHTML = "";

    const first = document.createElement("option");

    first.value = "";
    first.textContent = "Select an available time";
    first.selected = true;

    timeSelect.appendChild(first);

    let available = 0;

    timeSlots.forEach(([slotValue, label]) => {
      const option = document.createElement("option");

      option.value = slotValue;

      if (bookedSlots.has(slotKey(date, slotValue))) {
        option.textContent = `${label} — Reserved`;

        option.disabled = true;
      } else {
        option.textContent = label;
        available++;
      }

      timeSelect.appendChild(option);
    });

    if (available === 0) {
      resetTimeSelect("No times available for this date");
    } else {
      timeSelect.disabled = false;
    }
  }

  /* =========================================================
     SELECT DATE
  ========================================================= */

  function selectDate(date) {
    date = normalizeDate(date);

    if (!date) return;

    if (isSunday(date)) {
      showMessage(
        "The academy is closed on Sundays. Please choose another date.",
        "info",
      );

      return;
    }

    const selected = new Date(`${date}T12:00:00`);

    const today = new Date();

    today.setHours(0, 0, 0, 0);

    if (selected < today) {
      showMessage("Please choose a future lesson date.", "error");

      return;
    }

    dateInput.value = date;

    updateTimeOptions();

    clearMessage();

    if (window.innerWidth <= 700) {
      dateInput.closest(".form-section")?.scrollIntoView({
        behavior: "smooth",
        block: "center",
      });
    }
  }

  dateInput?.addEventListener("change", () => {
    selectDate(dateInput.value);

    if (calendar && dateInput.value) {
      calendar.gotoDate(dateInput.value);
    }
  });

  /* =========================================================
     CALENDAR EVENTS
  ========================================================= */

  async function getCalendarEvents() {
    try {
      const response = await fetch(API.bookings, {
        headers: {
          Accept: "application/json",
        },
        cache: "no-store",
      });

      if (!response.ok) {
        throw new Error(`HTTP ${response.status}`);
      }

      const data = await response.json();

      return Array.isArray(data) ? data : [];
    } catch (error) {
      console.error("Calendar events error:", error);

      return [];
    }
  }

  function convertEvents(items) {
    return items
      .map((item) => {
        if (!item) return null;

        let start = item.start;

        if (!start && item.lesson_date) {
          start = item.lesson_date;

          if (item.lesson_time) {
            start += `T${normalizeTime(item.lesson_time)}`;
          }
        }

        if (!start) return null;

        return {
          title: item.title || "Reserved",

          start,

          display: "block",

          backgroundColor: "#a9b6c2",

          borderColor: "#a9b6c2",

          textColor: "#ffffff",
        };
      })
      .filter(Boolean);
  }

  /* =========================================================
     INITIALIZE CALENDAR
  ========================================================= */

  async function initializeCalendar() {
    if (!calendarElement) {
      return;
    }

    if (typeof FullCalendar === "undefined") {
      showMessage(
        "The booking calendar could not load. Please refresh the page.",
        "error",
      );

      return;
    }

    calendar = new FullCalendar.Calendar(calendarElement, {
      initialView: "dayGridMonth",

      height: "auto",

      contentHeight: "auto",

      fixedWeekCount: false,

      firstDay: 1,

      selectable: true,

      selectMirror: true,

      dayMaxEvents: 2,

      headerToolbar: {
        left: "prev,next today",

        center: "title",

        right: "",
      },

      buttonText: {
        today: "Today",
      },

      events: convertEvents(await getCalendarEvents()),

      dateClick(info) {
        selectDate(info.dateStr);
      },

      select(info) {
        selectDate(info.startStr);

        calendar.unselect();
      },

      datesSet() {
        updateTimeOptions();
      },
    });

    calendar.render();
  }

  /* =========================================================
     REFRESH CALENDAR
  ========================================================= */

  async function refreshCalendar() {
    if (!calendar) return;

    const events = convertEvents(await getCalendarEvents());

    calendar.removeAllEvents();

    events.forEach((event) => {
      calendar.addEvent(event);
    });
  }

  /* =========================================================
     VALIDATION
  ========================================================= */

  function validEmail(email) {
    return /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email);
  }

  function validPhone(phone) {
    return String(phone || "").replace(/\D/g, "").length >= 7;
  }

  function validateForm() {
    clearMessage();

    if (!form.checkValidity()) {
      form.reportValidity();

      showMessage("Please complete all required fields.", "error");

      return false;
    }

    const required = [
      ["student_name", "Please enter the student's full name."],

      ["parent_name", "Please enter the parent or guardian name."],

      ["emergency_contact", "Please enter an emergency contact."],

      ["swimming_experience", "Please describe the swimmer's experience."],

      ["lesson_date", "Please select a lesson date."],

      ["lesson_time", "Please select an available lesson time."],

      ["lesson_type", "Please select a lesson type."],

      ["package", "Please select a package."],
    ];

    for (const [id, message] of required) {
      if (!value(id)) {
        showMessage(message, "error");

        document.getElementById(id)?.focus();

        return false;
      }
    }

    if (!validEmail(value("email"))) {
      showMessage("Please enter a valid email address.", "error");

      document.getElementById("email")?.focus();

      return false;
    }

    if (!validPhone(value("phone"))) {
      showMessage("Please enter a valid phone number.", "error");

      document.getElementById("phone")?.focus();

      return false;
    }

    if (!validPhone(value("emergency_phone"))) {
      showMessage("Please enter a valid emergency phone number.", "error");

      document.getElementById("emergency_phone")?.focus();

      return false;
    }

    if (isSunday(value("lesson_date"))) {
      showMessage(
        "The academy is closed on Sundays. Please select another date.",
        "error",
      );

      return false;
    }

    if (getSelectedPrice() === null) {
      showMessage("Please select a valid lesson type and package.", "error");

      return false;
    }

    if (agreement && !agreement.checked) {
      showMessage(
        "Please confirm that the registration information is accurate.",
        "error",
      );

      agreement.focus();

      return false;
    }

    if (bookedSlots.has(slotKey(value("lesson_date"), value("lesson_time")))) {
      updateTimeOptions();

      showMessage(
        "That time is already reserved. Please choose another time.",
        "error",
      );

      return false;
    }

    return true;
  }

  /* =========================================================
     BUILD BOOKING PAYLOAD
  ========================================================= */

  function buildPayload() {
    const type = value("lesson_type");

    const pkg = value("package");

    const price = lessonPrices[type]?.[pkg];

    return {
      name: value("student_name"),

      email: value("email"),

      phone: value("phone"),

      lesson_type: type,

      package: pkg,

      price: `$${Number(price).toFixed(2)}`,

      lesson_date: value("lesson_date"),

      lesson_time: value("lesson_time"),

      student_name: value("student_name"),

      dob: value("dob"),

      age: value("age"),

      parent_name: value("parent_name"),

      emergency_contact: value("emergency_contact"),

      emergency_phone: value("emergency_phone"),

      swimming_experience: value("swimming_experience"),

      medical: value("medical"),

      notes: value("notes"),

      payment_method: "pending_approval",

      payment_status: "pending",

      status: "pending",
    };
  }

  /* =========================================================
     SEND FOR APPROVAL
  ========================================================= */

  async function sendForApproval() {
    if (isSubmitting || !validateForm()) {
      return;
    }

    setLoading(true);

    showMessage(
      "Checking availability and sending your registration...",
      "info",
    );

    try {
      await loadBookedSlots();

      const date = value("lesson_date");

      const time = value("lesson_time");

      if (bookedSlots.has(slotKey(date, time))) {
        updateTimeOptions();

        throw new Error(
          "That time was just reserved. Please choose another available time.",
        );
      }

      const response = await fetch(API.createBooking, {
        method: "POST",

        headers: {
          "Content-Type": "application/json",

          Accept: "application/json",
        },

        body: JSON.stringify(buildPayload()),
      });

      let result = null;

      try {
        result = await response.json();
      } catch {
        result = null;
      }

      if (!response.ok || !result?.success) {
        throw new Error(
          result?.error ||
            result?.message ||
            `Registration failed (${response.status}).`,
        );
      }

      console.log("BOOKING CREATED:", result.booking_id);

      await loadBookedSlots();

      await refreshCalendar();

      form.reset();

      if (dateInput) {
        dateInput.value = "";
      }

      resetTimeSelect();

      updatePrice();

      if (calendar) {
        calendar.unselect();
      }

      openSuccessModal();
    } catch (error) {
      console.error("SEND FOR APPROVAL ERROR:", error);

      showMessage(
        error.message ||
          "We could not send your registration. Please try again.",
        "error",
      );
    } finally {
      setLoading(false);
    }
  }

  /* =========================================================
     SUCCESS MODAL
  ========================================================= */

  function openSuccessModal() {
    if (!successModal) {
      showMessage("Your registration was submitted successfully.", "success");

      return;
    }

    successModal.classList.add("is-open");

    successModal.setAttribute("aria-hidden", "false");

    document.body.style.overflow = "hidden";

    setTimeout(() => successDoneBtn?.focus(), 50);
  }

  function closeModal() {
    if (!successModal) return;

    successModal.classList.remove("is-open");

    successModal.setAttribute("aria-hidden", "true");

    document.body.style.overflow = "";
  }

  closeSuccessModal?.addEventListener("click", closeModal);

  successDoneBtn?.addEventListener("click", closeModal);

  document.querySelectorAll("[data-close-modal]").forEach((element) => {
    element.addEventListener("click", closeModal);
  });

  document.addEventListener("keydown", (event) => {
    if (event.key === "Escape") {
      closeModal();
    }
  });

  /* =========================================================
     PHONE FORMATTING
  ========================================================= */

  function formatPhone(input) {
    if (!input) return;

    input.addEventListener("input", () => {
      const digits = input.value.replace(/\D/g, "").slice(0, 10);

      if (digits.length <= 3) {
        input.value = digits;
      } else if (digits.length <= 6) {
        input.value = `(${digits.slice(0, 3)}) ${digits.slice(3)}`;
      } else {
        input.value = `(${digits.slice(0, 3)}) ${digits.slice(
          3,
          6,
        )}-${digits.slice(6)}`;
      }
    });
  }

  formatPhone(document.getElementById("phone"));

  formatPhone(document.getElementById("emergency_phone"));

  /* =========================================================
     AUTOMATIC AGE CALCULATION
  ========================================================= */

  const dob = document.getElementById("dob");

  const age = document.getElementById("age");

  dob?.addEventListener("change", () => {
    if (!dob.value || !age) {
      return;
    }

    const birthDate = new Date(`${dob.value}T12:00:00`);

    if (Number.isNaN(birthDate.getTime())) {
      return;
    }

    const today = new Date();

    let calculatedAge = today.getFullYear() - birthDate.getFullYear();

    const monthDifference = today.getMonth() - birthDate.getMonth();

    if (
      monthDifference < 0 ||
      (monthDifference === 0 && today.getDate() < birthDate.getDate())
    ) {
      calculatedAge--;
    }

    if (calculatedAge >= 0 && calculatedAge <= 120) {
      age.value = calculatedAge;
    }
  });

  /* =========================================================
     BUTTON EVENTS
  ========================================================= */

  sendApprovalBtn?.addEventListener("click", (event) => {
    event.preventDefault();

    sendForApproval();
  });

  form.addEventListener("submit", (event) => {
    event.preventDefault();

    sendForApproval();
  });

  /* =========================================================
     INITIALIZATION
  ========================================================= */

  resetTimeSelect();

  updatePrice();

  Promise.all([loadBookedSlots(), initializeCalendar()]).catch((error) => {
    console.error("BOOKING INITIALIZATION ERROR:", error);
  });

  /* =========================================================
     REFRESH AVAILABILITY EVERY 60 SECONDS
  ========================================================= */

  setInterval(async () => {
    if (document.visibilityState !== "visible" || isSubmitting) {
      return;
    }

    await loadBookedSlots();

    await refreshCalendar();
  }, 60000);
});
